sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaUnboundedReader.java

```diff
@@ -653,7 +664,8 @@ private void nextBatch() throws IOException {
       LOG.debug("Record count: " + records.count());
     }
 
-    partitionStates.forEach(p -> p.recordIter = records.records(p.topicPartition).iterator());
+    ConsumerRecords<byte[], byte[]> finalRecords = records;
```

Comment:
is the intention to create another reference to `records` here?
any reason why we can't just use `records`?

Reply:
Yes, it was intentional, since when I don't do it, I was getting the following compile error:

error: local variables referenced from a lambda expression must be final or effectively final
    partitionStates.forEach(p -> p.recordIter = records.records(p.topicPartition).iterator());

Though I don't quite understand why that is being thrown if the original expression was valid. (as nothing has functionally changed there)

Decided to rewrite the lambda and move that logic into a for loop instead.

Reply:
ah yea thats a thing with java lambdas sgtm